File: core/src/org/rexcellentgames/burningknight/entity/item/weapon/ConfettiGun.java
package org.rexcellentgames.burningknight.entity.item.weapon;

import org.rexcellentgames.burningknight.Dungeon;
import org.rexcellentgames.burningknight.entity.fx.Confetti;
import org.rexcellentgames.burningknight.entity.item.weapon.gun.Gun;
import org.rexcellentgames.burningknight.util.Random;
import org.rexcellentgames.burningknight.util.geometry.Point;

public class ConfettiGun extends Gun {
	{
		origin.x = 4;
		origin.y = 3;
		hole.x = 11;
		hole.y = 5;
		setAccuracy(20);
	}

	@Override
	protected void sendBullets() {
		Point aim = this.owner.getAim();
		float a = (float) (this.owner.getAngleTo(aim.x, aim.y) - Math.PI * 2);

		for (int i = 0; i < 10; i++) {
			float an = (float) (a + Math.toRadians(Random.newFloat(-this.getAccuracy(), this.getAccuracy())));

			Confetti fx = new Confetti();

			float x = this.owner.x + this.owner.w / 2;
			float y = this.owner.y + this.owner.h / 4 + region.getRegionHeight() / 2 - 2;

			fx.x = x + this.getAimX(0, 0);
			fx.y = y + this.getAimY(0, 0);

			float f = Random.newFloat(40, 80f);

			fx.vel.x = (float) Math.cos(an) * f;
			fx.vel.y = (float) Math.sin(an) * f;

			Dungeon.area.add(fx);
		}
	}
}

File: core/src/org/rexcellentgames/burningknight/entity/item/weapon/gun/ConfettiGun.kt
package org.rexcellentgames.burningknight.entity.item.weapon.gun

import org.rexcellentgames.burningknight.Dungeon
import org.rexcellentgames.burningknight.entity.fx.Confetti
import org.rexcellentgames.burningknight.game.input.Input
import org.rexcellentgames.burningknight.util.Random

class ConfettiGun : Gun() {
	init {
		origin.x = 4f
		origin.y = 3f
		hole.x = 11f
		hole.y = 5f
		accuracy = 20f
		damage = 4
		ammoMax = 1
		sprite = "item-tada_gun"
		reloadRate = 3f
	}

	override fun sendBullets() {
		var a = (this.owner.getAngleTo(Input.instance.worldMouse.x, Input.instance.worldMouse.y) - Math.PI * 2).toFloat()
		val ac = this.accuracy

		for (i in 0 .. (this.level * 2 + 1)) {
			this.sendBullet((a + Math.toRadians(Random.newFloat(-ac, ac).toDouble())).toFloat())
		}

		val aim = this.owner.aim
		a = (this.owner.getAngleTo(aim.x, aim.y) - Math.PI * 2).toFloat()

		for (i in 0..9) {
			val an = (a + Math.toRadians(Random.newFloat(-ac, ac).toDouble())).toFloat()

			val fx = Confetti()

			val x = this.owner.x + this.owner.w / 2
			val y = this.owner.y + this.owner.h / 4 + (region.regionHeight / 2).toFloat() - 2

			fx.x = x + this.getAimX(0f, 0f)
			fx.y = y + this.getAimY(0f, 0f)

			val f = Random.newFloat(40f, 80f) * 1.3f

			fx.vel.x = Math.cos(an.toDouble()).toFloat() * f
			fx.vel.y = Math.sin(an.toDouble()).toFloat() * f

			Dungeon.area.add(fx)
		}
	}

	override fun getMaxLevel(): Int {
		return 6
	}
}